File: tressette/graphical/ui.py
from __future__ import annotations

from typing import Callable

import tabletop.models as kt
from tabletop.config import tweak
from tabletop.rendering import (
    color_from_tuple,
    render_text,
    text_width,
)
from tabletop.ui import place_inside, place_next
from pyray import (
    Color,
    Rectangle,
    draw_rectangle_rounded,
    draw_rectangle_rounded_lines_ex,
)

from tressette.game.models import Game_State, Suit

# Italian suit short labels and tints — matches carte napoletane.
SUIT_LETTER = {
    Suit.COPPE: "Coppe",
    Suit.DENARI: "Denari",
    Suit.SPADE: "Spade",
    Suit.BASTONI: "Bastoni",
}

SUIT_COLOR = {
    Suit.COPPE: Color(180, 50, 70, 255),
    Suit.DENARI: Color(210, 170, 30, 255),
    Suit.SPADE: Color(70, 110, 190, 255),
    Suit.BASTONI: Color(80, 150, 80, 255),
}

# Rank labels: 1..7 numeric, then Donna / Cavallo / Re for face cards.
RANK_LABEL = {
    1: "1",
    2: "2",
    3: "3",
    4: "4",
    5: "5",
    6: "6",
    7: "7",
    8: "Donna",
    9: "Cavallo",
    10: "Re",
}


def make_tressette_stacks(both_hands_visible: bool) -> list[kt.Stack]:
    """Layout for the 6 zones on a Tressette table.

    Zones (in order, indices below match what agent_ui uses):
      0: p0_hand     — bottom player hand
      1: p1_hand     — top player hand
      2: p0_tricks   — bottom player's trick pile (face-down stack)
      3: p1_tricks   — top player's trick pile
      4: stock       — face-down draw pile, centered
      5: table       — current trick (0-2 cards face-up in the middle).
    """
    W = tweak["window_width"]
    H = tweak["window_height"]
    w = tweak["card_width"]
    h = tweak["card_height"]
    margin = 30

    spread_hand = int(w)
    spread_pile = -3

    window = Rectangle(0, 0, W, H)
    hand_width = spread_hand * 9 + w  # fits up to 10 cards.

    p0_hand = place_inside(
        window, hand_width, h, x="center", y="bottom", padding=margin
    )
    p1_hand = place_inside(window, hand_width, h, x="center", y="top", padding=margin)

    p0_tricks = place_next(p0_hand, w, h, x="right", y="center", padding=margin)
    p1_tricks = place_next(p1_hand, w, h, x="left", y="center", padding=margin)

    stock = place_inside(window, w, h, x="center", y="center", padding=0)
    stock.x -= int(w * 1.5)

    # Two card-slots side by side for the played trick (leader on the left).
    table = place_inside(window, 2 * w + 30, h, x="center", y="center", padding=0)
    table.x += int(w * 0.4)

    return [
        kt.Stack(
            p0_hand, spread_x=spread_hand, spread_y=0, face_up=True, name="p0_hand"
        ),
        kt.Stack(
            p1_hand,
            spread_x=spread_hand,
            spread_y=0,
            face_up=both_hands_visible,
            name="p1_hand",
        ),
        kt.Stack(
            p0_tricks, spread_x=0, spread_y=spread_pile, face_up=False, name="p0_tricks"
        ),
        kt.Stack(
            p1_tricks, spread_x=0, spread_y=spread_pile, face_up=False, name="p1_tricks"
        ),
        kt.Stack(stock, spread_x=0, spread_y=spread_pile, face_up=False, name="stock"),
        kt.Stack(table, spread_x=int(w + 30), spread_y=0, face_up=True, name="table"),
    ]


# Stack indices in the list returned above.
HAND = (0, 1)
TRICKS = (2, 3)
STOCK_IDX = 4
TABLE_IDX = 5


def make_card_draw_callback(state: Game_State, ui_state) -> Callable[[kt.Card], None]:
    """Render rank + suit on the card face. No card art."""

    def draw(card: kt.Card):
        c = state.all_cards[card.id]
        w = tweak["card_width"]
        h = tweak["card_height"]
        rank_text = RANK_LABEL[c.rank]
        suit_text = SUIT_LETTER[c.suit]
        col = SUIT_COLOR[c.suit]

        # Big rank number/word, centered horizontally near the top.
        rank_size = 56 if len(rank_text) == 1 else 32
        tw = text_width(rank_text, rank_size)
        render_text(rank_text, w // 2 - tw // 2, int(h * 0.18), rank_size, col)

        # Suit name underneath.
        suit_size = 22
        sw = text_width(suit_text, suit_size)
        render_text(suit_text, w // 2 - sw // 2, int(h * 0.55), suit_size, col)

        # Tiny rank in the bottom-right corner so overlapped fanned cards still
        # show their value.
        small_size = 22
        render_text(
            rank_text,
            w - text_width(rank_text, small_size) - 10,
            h - small_size - 10,
            small_size,
            col,
        )

        # Highlight border for legal cards (driven by ui_state.highlighted_cards).
        if card.id in ui_state.highlighted_cards.values():
            draw_rectangle_rounded_lines_ex(
                Rectangle(0, 0, w, h),
                0.18,
                8,
                4,
                color_from_tuple(tweak["highlight_color"]),
            )

    return draw


def draw_player_hud(player_index: int, score: int, is_current: bool, hud_y: int):
    label = f"Player {player_index + 1}: {score}"
    color = Color(200, 200, 200, 255) if is_current else Color(120, 120, 120, 200)
    render_text(label, 30, hud_y, 28, color)


def draw_game_over_overlay(scores: list[int]):
    from pyray import draw_rectangle, get_screen_height, get_screen_width

    W = get_screen_width()
    H = get_screen_height()
    draw_rectangle(0, 0, W, H, color_from_tuple(tweak["modal_overlay"]))
    title = "GAME OVER"
    if scores[0] > scores[1]:
        msg = "Player 1 wins!"
    elif scores[1] > scores[0]:
        msg = "Player 2 wins!"
    else:
        msg = "It's a tie."
    score_line = f"{scores[0]} - {scores[1]}"
    render_text(
        title, W // 2 - text_width(title, 60) // 2, 320, 60, Color(255, 255, 255, 255)
    )
    render_text(
        msg, W // 2 - text_width(msg, 36) // 2, 410, 36, Color(255, 215, 0, 255)
    )
    render_text(
        score_line,
        W // 2 - text_width(score_line, 30) // 2,
        470,
        30,
        Color(200, 200, 200, 255),
    )

File: tressette/graphical/ui.cpp
#include "ui.h"

#include <tabletop/config.h>
#include <tabletop/rendering.h>

#include <string>

// Suit display helpers.

static const char* suit_name(tressette::Suit s) {
  switch (s) {
    case tressette::Suit::COPPE:   return "Coppe";
    case tressette::Suit::DENARI:  return "Denari";
    case tressette::Suit::SPADE:   return "Spade";
    case tressette::Suit::BASTONI: return "Bastoni";
  }
  return "";
}

static Color suit_color(tressette::Suit s) {
  switch (s) {
    case tressette::Suit::COPPE:   return Color{180, 50,  70,  255};
    case tressette::Suit::DENARI:  return Color{210, 170, 30,  255};
    case tressette::Suit::SPADE:   return Color{70,  110, 190, 255};
    case tressette::Suit::BASTONI: return Color{80,  150, 80,  255};
  }
  return Color{0, 0, 0, 255};
}

// Rank labels: 1-7 numeric, 8-10 face-card names.
static const char* rank_label(int rank) {
  switch (rank) {
    case 1:  return "1";
    case 2:  return "2";
    case 3:  return "3";
    case 4:  return "4";
    case 5:  return "5";
    case 6:  return "6";
    case 7:  return "7";
    case 8:  return "Donna";
    case 9:  return "Cavallo";
    case 10: return "Re";
  }
  return "?";
}

std::vector<Thing> make_tressette_stacks(bool both_hands_visible) {
  const int W          = tt::WINDOW_WIDTH;
  const int H          = tt::WINDOW_HEIGHT;
  const int w          = tt::CARD_WIDTH;
  const int h          = tt::CARD_HEIGHT;
  const int margin     = 30;
  const int spread_hand = w;
  const int spread_pile = -3;
  const int hand_width = spread_hand * 9 + w;  // fits up to 10 cards.

  Rectangle window = {0.0f, 0.0f, (float)W, (float)H};

  Rectangle p0_hand_r   = place_inside(window, hand_width, h, "center", "bottom", margin);
  Rectangle p1_hand_r   = place_inside(window, hand_width, h, "center", "top",    margin);
  Rectangle p0_tricks_r = place_next(p0_hand_r, w, h, "right", "center", margin);
  Rectangle p1_tricks_r = place_next(p1_hand_r, w, h, "left",  "center", margin);
  Rectangle stock_r     = place_inside(window, w, h, "center", "center", 0);
  stock_r.x -= (float)(w * 3 / 2);
  Rectangle table_r     = place_inside(window, 2 * w + 30, h, "center", "center", 0);
  table_r.x += (float)(w * 2 / 5);

  auto make = [](Rectangle r, float sx, float sy, bool fu, const char* name) {
    Thing t;
    t.rect     = r;
    t.spread_x = sx;
    t.spread_y = sy;
    t.face_up  = fu;
    t.name     = name;
    return t;
  };

  return {
    make(p0_hand_r,   (float)spread_hand, 0.0f,            true,               "p0_hand"),
    make(p1_hand_r,   (float)spread_hand, 0.0f,            both_hands_visible, "p1_hand"),
    make(p0_tricks_r, 0.0f,               (float)spread_pile, false,           "p0_tricks"),
    make(p1_tricks_r, 0.0f,               (float)spread_pile, false,           "p1_tricks"),
    make(stock_r,     0.0f,               (float)spread_pile, false,           "stock"),
    make(table_r,     (float)(w + 30),    0.0f,            true,               "table"),
  };
}

std::function<void(Thing&)> make_card_draw_callback(
  const tressette::Game_State& state, UI_State& ui_state
) {
  return [&state, &ui_state](Thing& thing) {
    const tressette::Card& c    = state.all_cards[thing.id];
    const char*            rlbl = rank_label(c.rank);
    const char*            slbl = suit_name(c.suit);
    Color                  col  = suit_color(c.suit);
    int                    w    = tt::CARD_WIDTH;
    int                    h    = tt::CARD_HEIGHT;

    // Big rank number/word near the top, horizontally centered.
    int rank_size = ((int)std::string(rlbl).size() == 1) ? 56 : 32;
    int tw        = text_width(rlbl, rank_size);
    render_text(rlbl, (float)(w / 2 - tw / 2), h * 0.18f, rank_size, col);

    // Suit name underneath.
    int suit_size = 22;
    int sw        = text_width(slbl, suit_size);
    render_text(slbl, (float)(w / 2 - sw / 2), h * 0.55f, suit_size, col);

    // Tiny rank in the bottom-right corner so fanned cards still show value.
    int small_size = 22;
    int rw         = text_width(rlbl, small_size);
    render_text(rlbl, (float)(w - rw - 10), (float)(h - small_size - 10), small_size, col);

    // Highlight border for legal cards.
    if (ui_state.highlighted_cards.count(thing.id) > 0) {
      DrawRectangleRoundedLinesEx(
        Rectangle{0.0f, 0.0f, (float)w, (float)h},
        0.18f, 8, 4.0f,
        Color{255, 215, 0, 200}
      );
    }
  };
}

void draw_tressette_player_hud(
  int player_index, int score, bool is_current, int hud_y
) {
  std::string label =
    "Player " + std::to_string(player_index + 1) + ": " + std::to_string(score);
  Color col = is_current ? Color{200, 200, 200, 255} : Color{120, 120, 120, 200};
  render_text(label, 30.0f, (float)hud_y, 28, col);
}

void draw_tressette_game_over_screen(
  Table_State& table_state, const std::vector<int>& scores
) {
  const int   W          = tt::WINDOW_WIDTH;
  const int   H          = tt::WINDOW_HEIGHT;
  const char* title      = "GAME OVER";
  const char* msg =
    (scores[0] > scores[1]) ? "Player 1 wins!" :
    (scores[1] > scores[0]) ? "Player 2 wins!" : "It's a tie.";
  std::string score_line =
    std::to_string(scores[0]) + " - " + std::to_string(scores[1]);

  while (!WindowShouldClose()) {
    BeginDrawing();
    draw_background(0.0f);
    draw_table(table_state);
    DrawRectangle(0, 0, W, H, Color{0, 0, 0, 160});
    render_text(title,      (float)(W / 2 - text_width(title,      60) / 2), 320.0f, 60, Color{255, 255, 255, 255});
    render_text(msg,        (float)(W / 2 - text_width(msg,        36) / 2), 410.0f, 36, Color{255, 215, 0,   255});
    render_text(score_line, (float)(W / 2 - text_width(score_line, 30) / 2), 470.0f, 30, Color{200, 200, 200, 255});
    EndDrawing();
  }
}
